lib/AST/ASTImporter.cpp

```diff
@@ -2559,7 +2545,18 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
     }
 
   // Import the body, if any.
-  if (Stmt *FromBody = D->getBody()) {
+  const FunctionDecl* FromDefinition = nullptr;
+  if (Stmt *FromBody = D->getBody(FromDefinition)) {
+
+    // If we have the defintion in the FromCtx, mark that as imported.
+    // This handles difficulties when we have a recursive function with a
+    // prototype and a definition in the same ToCtx.  E.g. importing the
+    // prototype below would import the definition twice without this.
+    // "void f(); void f() { f(); }"
+    // This is because in the definition the CallExpression refers to the
+    // definition, not to the prototype.
+    Importer.Imported(const_cast<FunctionDecl*>(FromDefinition), ToFunction);
```

Comment:
This line introduces the following assert (and it is very frequent):
```
clang-6.0: ../tools/clang/lib/AST/ASTImporter.cpp:7465: virtual clang::Decl* clang::ASTImporter::Imported(clang::Decl*, clang::Decl*): Assertion `(Pos == ImportedDecls.end() || Pos->second == To) && "Try to import an already imported Decl"' failed.
#0 0x00007f01a28b001a llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/egborho/OSLLVM/build/../lib/Support/Unix/Signals.inc:402:0
#1 0x00007f01a28addbe llvm::sys::RunSignalHandlers() /home/egborho/OSLLVM/build/../lib/Support/Signals.cpp:50:0
#2 0x00007f01a28adf32 SignalHandler(int) /home/egborho/OSLLVM/build/../lib/Support/Unix/Signals.inc:242:0
#3 0x00007f01a12574b0 (/lib/x86_64-linux-gnu/libc.so.6+0x354b0)
#4 0x00007f01a1257428 gsignal /build/glibc-Cl5G7W/glibc-2.23/signal/../sysdeps/unix/sysv/linux/raise.c:54:0
#5 0x00007f01a125902a abort /build/glibc-Cl5G7W/glibc-2.23/stdlib/abort.c:91:0
#6 0x00007f01a124fbd7 __assert_fail_base /build/glibc-Cl5G7W/glibc-2.23/assert/assert.c:92:0
#7 0x00007f01a124fc82 (/lib/x86_64-linux-gnu/libc.so.6+0x2dc82)
#8 0x00007f019fa42b40 clang::ASTImporter::Imported(clang::Decl*, clang::Decl*) /home/egborho/OSLLVM/build/../tools/clang/lib/AST/ASTImporter.cpp:7479:0
#9 0x00007f019fa454d7 clang::ASTNodeImporter::VisitFunctionDecl(clang::FunctionDecl*) /home/egborho/OSLLVM/build/../tools/clang/lib/AST/ASTImporter.cpp:2575:0
#10 0x00007f019fa45d0f clang::ASTImporter::Import(clang::Decl*) /home/egborho/OSLLVM/build/../tools/clang/lib/AST/ASTImporter.cpp:6826:0
```

Maybe we should first check if we already imported the function definition and only attempt to import it if we did not import it yet?

Reply:
Actually, the if which Balazs suggests would solve this. That way we would first check if we already imported the function definition and only attempt to **mark that as imported** if we did not import it yet?

Reply:
This means the syntax tree will be manipulated to some extent (this was the case already). The actual definition of the function can be merged with other declarations. (The `FromDefinition` node will not imported any more because it is marked already as imported without really importing it.)

Reply:
Yes, that is right. I believe this is because we don't want to have redundant nodes in the AST. Imagine a TU with a prototype and another TU with the same prototype; if we import from both we still want to have only one prototype. If a definition is available, then we import the definition immediately. So the AST will not be the same as literally merging the ASTs. I think the current `ImportFunctions` test cases demonstrate it quite well how do want to eliminate redundancies.